README.md
# tiy-homework-string-katas-clojure

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-08-11
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-08-11
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/tiy-homework-string-katas-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/tiy-homework-string-katas-clojure/compare/0.1.0...0.1.1


test/tiy_homework_string_katas_clojure/core_test.clj
(ns tiy-homework-string-katas-clojure.core-test
  (:require [clojure.test :refer :all]
            [tiy-homework-string-katas-clojure.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/tiy_homework_string_katas_clojure/katas.clj
(ns tiy-homework-string-katas-clojure.katas
  (:require [clojure.string :as s]))
;  (:require [clojure.math.numeric-tower :as math]))


;The parameter weekday is true if it is a weekday, and the parameter vacation is
;true if we are on vacation. We sleep in if it is not a weekday or we're on
;vacation. Return true if we sleep in.
(defn sleep-in [week-day vacation]
  (or (not week-day)  vacation))

;We have two monkeys, a and b, and the parameters aSmile and bSmile indicate if
;each is smiling. We are in trouble if they are both smiling or if neither of
;them is smiling. Return true if we are in trouble.
(defn monkey-trouble [aSmile, bSmile]
  (or (and aSmile bSmile) (not (and aSmile bSmile))))

;Given two int values, return their sum. Unless the two values are the
;same, then return double their sum.
(defn sum-double [a, b]
  (if (= a b)
    (* 2 (+ a b))
    (+ a b)))


;Given an int n, return the absolute difference between n and 21, except
;return double the absolute difference if n is over 21.
(defn diff21 [n]
  (if (> n  21)
    (* 2 (- n 21))
    (- 21 n)))

;We have a loud talking parrot. The "hour" parameter is the current hour
;time in the range 0..23. We are in trouble if the parrot is talking
;and the hour is before 7 or after 20. Return true if we are in trouble.
(defn parrot-trouble [talking hour]
  (and talking (or (< hour 7) (> hour 20))))

;Given 2 ints, a and b, return true if one if them is 10 or if their sum is 10.
(defn makes10 [a b]
  (or (= a 10) (= b 10) (= (+ a b) 10)))


;Given an int n, return true if it is within 10 of 100 or 200.
;Note: Math.abs(num) computes the absolute value of a number.
(defn abs [n]
  (max n (- n)))

(defn near-hundred [n]
  (or (<= (abs (- 100 n)) 10) (<= (abs (- 200 n)) 10)))

;Given 2 int values, return true if one is negative and one is positive. Except
;if the parameter "negative" is true, then return true only if both are negative.
(defn pos-neg [a b negative]
  (and (not negative)
       (or
         (and (< a 0) (> b 0))
         (and (> a 0) (< b 0)))))


;Given a string, return a new string where "not " has been added to the front.
;However, if the string already begins with "not", return the string unchanged.
;Note: use .equals() to compare 2 strings.
(defn not-string [str-test]
  (if (<= (count str-test) 3)
    (str "not " str-test)
    (if (not (= (subs str-test 0 3) "not"))
     (str "not " str-test)
     str-test)))


(defn missing-char [str-test n]
  (if (< (count str-test) n)
    ("error!")
    (if (= n 0)
      (subs str-test (+ n 1));
      (if (= n (- (count str-test) 1))
        (subs str-test 0 (+ n 1))
        (str (subs str-test 0 n) (subs str-test (+ n 1)))))))

;Given a string, return a new string where the first and last chars have been exchanged.
(defn front-back [str-test]
  (if (<= (count str-test) 1)
    (str)
    (str (.charAt str-test (- (count str-test) 1)) (subs str-test 1 (- (count str-test) 1))  (.charAt str-test 0))))


;Given a string, we'll say that the front is the first 3 chars of the string. If the
;string length is less than 3, the front is whatever is there. Return a new string
;which is 3 copies of the front.
(defn front3 [str-test]
  (if (< (count str-test) 3)
    (str str-test str-test str-test)
    (let [first3 (subs str-test 0 3)]
      (str first3 first3 first3))))

;Given a string, take the last char and return a new string with the last char added
;at the front and back, so "cat" yields "tcatt". The original string will be length 1 or more.*/
(defn back-around [str-test]
  (let [last  (.charAt str-test (- (count str-test) 1))]
     (str last str-test last)))


;Return true if the given non-negative number is a multiple of 3 or a multiple of 5.
;Use the % "mod" operator
(defn or35 [n]
  (or (= (mod n  3) 0) (= (mod n  5) 0)))


;Given a string, take the first 2 chars and return the string with the 2 chars added
;at both the front and back, so "kitten" yields"kikittenki". If the string length is
;less than 2, use whatever chars are there.*/
(defn front22 [str-test]
  (if (<= (count str-test) 2)
    (str str-test str-test str-test)
    (str (subs str-test 0,2) str-test (subs str-test 0,2))))

;Given a string, return true if the string starts with "hi" and false otherwise.
(defn start-hi [str-test]
   (if (< (count str-test) 2)
     false
     (if (= (subs str-test 0 2) "hi")
       true
       false)))

;Given two temperatures, return true if one is less than 0 and the other is greater than 100.
(defn icy-hot [temp1 temp2]
  (or (and (> temp1 100) (< temp2 0)) (and (< temp1 0) (> temp2 100))))

;Given 2 int values, return true if either of them is in the range 10..20 inclusive.
(defn in1020 [a b]
  (or (and (< a 21) (> a 9)) (and (> b 9) (< b 21))))


;We'll say that a number is "teen" if it is in the range 13..19 inclusive.
;Given 3 int values, return true if 1 or more of them are teen.
(defn has-teen [a b c]
  (or (and (>= a 13) (<= a 19))
    (and (>= b 13) (<= b 19))
    (and (>= c 13) (<= c 19))))


;We'll say that a number is "teen" if it is in the range 13..19 inclusive.
;Given 2 int values, return true if one or the other is teen, but not both.
(defn lone-teen [a b]
  (and (or (and (>= a 13) (<= a 19)) (>= b 13) (<= b 19))
      (not (and (>= a 13) (<= a 19) (>= b 13) (<= b 19)))))


;Given a string, if the string "del" appears starting at index 1,
;return a string where that "del" has been deleted. Otherwise, return the
;string unchanged.
(defn del-del [str-test]
  (if (and (> (count str-test) 3) (= (subs str-test 1 4) "del"))
    (str (subs str-test 0 1) (subs str-test 4))
    str-test))


;Return true if the given string begins with "mix", except the 'm'
;can be anything, so "pix", "9ix" .. all count.
(defn mix-start [str-test]
  (and (>= (count str-test) 3) (= (subs str-test 1, 3) "ix")))


;Given a string, return a string made of the first 2 chars (if present),
;however include first char only if it is 'o' and include the second only
;if it is 'z', so "ozymandias" yields "oz".
(defn start-oz [str-test]
  (if (>= (count str-test) 2)
    (let [o (subs str-test 0 1)
          z (subs str-test 1 2)]
      (if (and (= o "o") (= z "z"))
        (str o z)
        (if (= o "o")
          (str o)
          (if (= z "z")
            (str z)
            (str "")))))
    (if (= str-test "o")
      str-test
      "")))



;Given three int values, a b c, return the largest.
(defn int-max [a b c]
  (max a b c))


;Given 2 int values, return whichever value is nearest to the value 10,
;or return 0 in the event of a tie. Note that Math.abs(n) returns the
;absolute value of a number.
(defn close10 [a b]
  (if (= (abs (- 10 a)) (abs (- 10 b)))
    0
    (if ( > (abs (- 10 a)) (abs (- 10 b)))
      b
      a)))


;Given 2 int values, return true if they are both in the range 30..40
;'inclusive, or they are both in the range 40..50 inclusive.
(defn in3050 [a b]
  (or (and (>= a 30) (<= a 40) (>= b 30) (<= b 40))
      (and (>= a 40) (<= a 50) (>= b 40) (<= b 50))))


;Given 2 positive int values, return the larger value that is in the
;range 10..20 inclusive, or return 0 if neither is in that range.
(defn max1020 [a b]
  (if (and (>= a 10) (<= a 20) (>= b 10) (<= b 20))
    (if (> a b)
      a
      b)
    (if (and (>= a 10) (<= a 20))
      a
      (if (and (>= b 10) (<= b 20))
        b
        0))))


;Return true if the given string contains between 1 and 3 'e' chars.
(defn string-e [str-test]
  (if (.contains str-test "e")
    (<= (get (frequencies str-test) \e) 3)
    false))

;Given two non-negative int values, return true if they have the same last
;digit, such as with 27 and 57. Note that the % "mod" operator computes
;remainders, so 17 % 10 is 7.
(defn last-digit [a b]
  (= (mod a 10) (mod b 10)))


;Given a string, return a new string where the last 3 chars are now in upper
;case. If the string has less than 3 chars, uppercase whatever is there. Note
;that str-test.toUpperCase() returns the uppercase version of a string.
(defn end-up [str-test]
  (if (< (count str-test) 4)
    (s/upper-case str-test)
    (let [last3 (subs str-test (- (count str-test) 3))]
        (str (subs str-test 0 (- (count str-test) 3)) (s/upper-case last3)))))

;Given a non-empty string and an int N, return the string made starting with
;char 0, and then every Nth char of the string. So if N is 3, use char 0, 3,
;6, ... and so on. N is 1 or more.
(defn every-nth [str-test n]
  (for [i (range 0 (count str-test) n)]
       (str (subs str-test i (+ i 1)))))


;Given a string and a non-negative int n, return a larger string that is
;n copies of the original string.
(defn string-times [str-test cnt]
;  (def buffer str-test)
  (for [x (range 0 cnt)]
    (str str-test)))



(defn explode [str-test]
  (if (< (count str-test) 2)
    (str str-test)
    (for [i (range 0 (count str-test))]
         (str (nth str-test i)  " "))))


src/tiy_homework_string_katas_clojure/core.clj
(ns tiy-homework-string-katas-clojure.core
  (:use tiy-homework-string-katas-clojure.katas))

(defn -main []
  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (sleep-in)")
  (println "(sleep-in not weekday not vacation)?" (sleep-in false false))
  (println "(sleep-in weekday not vacation)?" (sleep-in true false))
  (println "(sleep-in not weekday vacation)?" (sleep-in false true))
  (println "(sleep-in not weekday vacation)?" (sleep-in true true))

  (println);
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing monkey-trouble()")
  (println "monkey-trouble true true: " (monkey-trouble true true))
  (println "monkey-trouble false false: " (monkey-trouble false false))
  (println "monkey-trouble true false: " (monkey-trouble true false))
  (println "monkey-trouble false true: " (monkey-trouble false true))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (sum-double)")
  (println "(sum-double 1 2)" (sum-double 1 2))
  (println "(sum-double 3 2)" (sum-double 3 2))
  (println "(sum-double 2 2)" (sum-double 2 2))
  (println "(sum-double -1 0)" (sum-double -1 0))
  (println "(sum-double 3 3)" (sum-double 3 3))
  (println "(sum-double 0 0)" (sum-double 0 0))
  (println "(sum-double 0 1)" (sum-double 0 1))
  (println "(sum-double 3 4)" (sum-double 3 4))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (diff21)")
  (println "(diff21 19)" (diff21 19))
  (println "(diff21 10)" (diff21 10))
  (println "(diff21 21)" (diff21 21))
  (println "(diff21 22)" (diff21 22))
  (println "(diff21 25)" (diff21 25))
  (println "(diff21 30)" (diff21 30))
  (println "(diff21 0)" (diff21 0))
  (println "(diff21 1)" (diff21 1))
  (println "(diff21 2)" (diff21 2))
  (println "(diff21 -1)" (diff21 -1))
  (println "(diff21 -2)" (diff21 -2))
  (println "(diff21 50)" (diff21 50))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing parrot-trouble()")
  (println "(parrot-trouble true 6): " (parrot-trouble true 6))
  (println "(parrot-trouble true 7): " (parrot-trouble true 7))
  (println "(parrot-trouble false 6): " (parrot-trouble false 6))
  (println "(parrot-trouble true 21): " (parrot-trouble true 21))
  (println "(parrot-trouble false 21): " (parrot-trouble false 21))
  (println "(parrot-trouble false 20): " (parrot-trouble false 20))
  (println "(parrot-trouble true 23): " (parrot-trouble true 23))
  (println "(parrot-trouble false 23): " (parrot-trouble false 23))
  (println "(parrot-trouble true 20): " (parrot-trouble true 20))
  (println "(parrot-trouble false 12): " (parrot-trouble false 12))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (makes10)")
  (println "(makes10 9 10): " (makes10 9 10))
  (println "(makes10 9 9): " (makes10 9 9))
  (println "(makes10 1 10): " (makes10 1 9))
  (println "(makes10 10 1): " (makes10 10 1))
  (println "(makes10 10 10): " (makes10 10 10))
  (println "(makes10 8 2): " (makes10 8 2))
  (println "(makes10 8 3): " (makes10 8 3))
  (println "(makes10 10 42): " (makes10 10 42))
  (println "(makes10 12 -2): " (makes10 12 -2))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (nearHundred)")
  (println "(near-hundred 93): " (near-hundred 93))
  (println "(near-hundred 90): " (near-hundred 90))
  (println "(near-hundred 89): " (near-hundred 89))
  (println "(near-hundred 110): " (near-hundred 110))
  (println "(near-hundred 111): " (near-hundred 111))
  (println "(near-hundred 121): " (near-hundred 121))
  (println "(near-hundred 0): " (near-hundred 0))
  (println "(near-hundred 5): " (near-hundred 5))
  (println "(near-hundred 191): " (near-hundred 191))
  (println "(near-hundred 189): " (near-hundred 189))
  (println "(near-hundred 190): " (near-hundred 190))
  (println "(near-hundred 200): " (near-hundred 200))
  (println "(near-hundred 210): " (near-hundred 210))
  (println "(near-hundred 221): " (near-hundred 211))
  (println "(near-hundred 290): " (near-hundred 290))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (pos-neg)")
  (println "(pos-neg 1 -1 false): " (pos-neg 1 -1 false))
  (println "(pos-neg -1 1 false): " (pos-neg -1 1 false))
  (println "(pos-neg -4 -5 true): " (pos-neg -4 -5 true))
  (println "(pos-neg -4 -5 false): " (pos-neg -4 -5 false))
  (println "(pos-neg -4 5 false): " (pos-neg -4 5 false))
  (println "(pos-neg -4 5 true): " (pos-neg -4 5 true))
  (println "(pos-neg 1 1 false): " (pos-neg 1 1 false))
  (println "(pos-neg -1 -1 false): " (pos-neg -1 -1 false))
  (println "(pos-neg 1 -1 true): " (pos-neg 1 -1 true))
  (println "(pos-neg -1 1 true): " (pos-neg -1 1 true))
  (println "(pos-neg 1 1 true): " (pos-neg 1 1 true))
  (println "(pos-neg -1 -1 true): " (pos-neg -1 -1 true))
  (println "(pos-neg 5 -5 false): " (pos-neg 5 -5 false))
  (println "(pos-neg -6 6 false): " (pos-neg -6 6 false))
  (println "(pos-neg -5 -6 false): " (pos-neg -5 -6 false))
  (println "(pos-neg -2 -1 false): " (pos-neg -2 -1 false))
  (println "(pos-neg 1 2 false): " (pos-neg 1 2 false))
  (println "(pos-neg -5 6 true): " (pos-neg -5 6 true))
  (println "(pos-neg -5 -5 true): " (pos-neg -5 -5 true))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (not-string)")
  (println "(not-string \"not candy\"): " (not-string "candy"))
  (println "(not-string \"X\"): " (not-string "x"))
  (println "(not-string \"not bad\"): " (not-string "not bad"))
  (println "(not-string \"bad\"): " (not-string "bad"))
  (println "(not-string \"not\"): " (not-string "not"))
  (println "(not-string \"is not\"): " (not-string "is not"))
  (println "(not-string \"no\"): " (not-string "no"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (missing-char)")
  (println "(missing-char \"kitten\" 1): " (missing-char "kitten" 1))
  (println "(missing-char \"kitten\" 0): " (missing-char "kitten" 0))
  (println "(missing-char \"kitten\" 4): " (missing-char "kitten" 4))
  (println "(missing-char \"Hi\" 0): " (missing-char "Hi" 0))
  (println "(missing-char \"Hi\" 1): " (missing-char "Hi" 1))
  (println "(missing-char \"code\" 0): " (missing-char "code" 0))
  (println "(missing-char \"code\" 1): " (missing-char "code" 1))
  (println "(missing-char \"code\" 2): " (missing-char "code" 2))
  (println "(missing-char \"code\" 3): " (missing-char "code" 3))
  (println "(missing-char \"chocolate\" 8): " (missing-char "chocolate" 8))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (front-back )")
  (println "(front-back \"code\"): " (front-back "code"))
  (println "(front-back \"a\"): " (front-back "a"))
  (println "(front-back \"ab\"): " (front-back "ab"))
  (println "(front-back \"abc\"): " (front-back "abc"))
  (println "(front-back \"\"): " (front-back ""))
  (println "(front-back \"Chocolate\"): " (front-back "Chocolate"))
  (println "(front-back \"aavJ\"): " (front-back "aavJ"))
  (println "(front-back \"hello\"): " (front-back "hello"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (front3)")
  (println "(front3 \"Java\"): " (front3 "Java"))
  (println "(front3 \"Chocolate\"): " (front3 "Chocolate"))
  (println "(front3 \"abc\"): " (front3 "abc"))
  (println "(front3 \"abcXYZ\"): " (front3 "abcXYZ"))
  (println "(front3 \"ab\"): " (front3 "ab"))
  (println "(front3 \"a\"): " (front3 "a"))
  (println "(front3 \"\"): " (front3 ""))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (back-around)")
  (println "(back-around \"cat\"): " (back-around "cat"))
  (println "(back-around \"Hello\"): " (back-around "Hello"))
  (println "(back-around \"a\"): " (back-around "a"))
  (println "(back-around \"abc\"): " (back-around "abc"))
  (println "(back-around \"read\"): " (back-around "read"))
  (println "(back-around \"boo\"): " (back-around "boo"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (or35)")
  (println "(or35 3): " (or35 3))
  (println "(or35 10): " (or35 10))
  (println "(or35 8): " (or35 8))
  (println "(or35 15): " (or35 15))
  (println "(or35 5): " (or35 5))
  (println "(or35 9): " (or35 9))
  (println "(or35 4): " (or35 4))
  (println "(or35 7): " (or35 7))
  (println "(or35 6): " (or35 6))
  (println "(or35 17): " (or35 17))
  (println "(or35 18): " (or35 18))
  (println "(or35 29): " (or35 29))
  (println "(or35 209): " (or35 20))
  (println "(or35 21): " (or35 21))
  (println "(or35 22): " (or35 22))
  (println "(or35 45): " (or35 45))
  (println "(or35 99): " (or35 99))
  (println "(or35 100): " (or35 100))
  (println "(or35 101): " (or35 101))
  (println "(or35 121): " (or35 121))
  (println "(or35 122): " (or35 122))
  (println "(or35 123): " (or35 123))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (front22)")
  (println "(front22 \"kitten\"): " (front22 "kitten"))
  (println "(front22 \"Ha\"): " (front22 "Ha"))
  (println "(front22 \"abc\"): " (front22 "abc"))
  (println "(front22 \"ab\"): " (front22 "ab"))
  (println "(front22 \"a\"): " (front22 "a"))
  (println "(front22 \"\"): " (front22 ""))
  (println "(front22 \"Logic\"): " (front22 "Logic"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (start-hi)")
  (println "(start-hi\"hi there\"): " (start-hi "hi there"))
  (println "(start-hi\"hi\"): " (start-hi "hi"))
  (println "(start-hi\"hello hi\"): " (start-hi "hello hi"))
  (println "(start-hi\"he\"): " (start-hi "he"))
  (println "(start-hi\"h\"): " (start-hi "h"))
  (println "(start-hi\"\"): " (start-hi ""))
  (println "(start-hi\"ho hi\"): " (start-hi "ho hi"))
  (println "(start-hi\"hi ho\"): " (start-hi "hi ho"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (icy-hot)")
  (println "(icy-hot 120 -1): " (icy-hot 120 -1))
  (println "(icy-hot -1 120): " (icy-hot -1 120))
  (println "(icy-hot 2 120): " (icy-hot 2 120))
  (println "(icy-hot -1 100): " (icy-hot -1 100))
  (println "(icy-hot -2 -2): " (icy-hot -2 -2))
  (println "(icy-hot 120 120): " (icy-hot 120 120))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing in1020()")
  (println "(in1020 12 99): " (in1020 12 99))
  (println "(in1020 21 12): " (in1020 21 12))
  (println "(in1020 8 99): " (in1020 8 99))
  (println "(in1020 99 10): " (in1020 99 10))
  (println "(in1020 20 20): " (in1020 20 20))
  (println "(in1020 21 21): " (in1020 21 21))
  (println "(in1020 9 9): " (in1020 9 9))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing has-teen()")
  (println "(has-teen 13 20 10): " (has-teen 13 20 10))
  (println "(has-teen 20 19 10): " (has-teen 20 19 10))
  (println "(has-teen 20 10 13): " (has-teen 20 10 13))
  (println "(has-teen 1 20 12): " (has-teen 1 20 12))
  (println "(has-teen 19 20 12): " (has-teen 19 20 12))
  (println "(has-teen 12 20 19): " (has-teen 12 20 19))
  (println "(has-teen 12 9 20): " (has-teen 12 9 20))
  (println "(has-teen 12 18 20): " (has-teen 12 18 20))
  (println "(has-teen 14 2 20): " (has-teen 14 2 20))
  (println "(has-teen 4 2 20): " (has-teen 4 2 20))
  (println "(has-teen 11 22 22): " (has-teen 11 22 22))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing loneTeen()")
  (println "(lone-teen 13 99): " (lone-teen 13 99))
  (println "(lone-teen 21 19): " (lone-teen 21 19))
  (println "(lone-teen 13 13): " (lone-teen 13 13))
  (println "(lone-teen 14 20): " (lone-teen 14 20))
  (println "(lone-teen 20 15): " (lone-teen 20 15))
  (println "(lone-teen 16 17): " (lone-teen 16 17))
  (println "(lone-teen 16 9): " (lone-teen 16 9))
  (println "(lone-teen 16 18): " (lone-teen 16 18))
  (println "(lone-teen 13 19): " (lone-teen 13 19))
  (println "(lone-teen 13 20): " (lone-teen 13 20))
  (println "(lone-teen 6 18): " (lone-teen 6 18))
  (println "(lone-teen 99 13): " (lone-teen 99 13))
  (println "(lone-teen 99 99): " (lone-teen 99 99))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (del-del)")
  (println "(del-del \"adelbc\")): " (del-del  "adelbc"))
  (println "(del-del \"adelHello\")): " (del-del  "adelHello"))
  (println "(del-del \"adedbc\"): " (del-del  "adedbc"))
  (println "(del-del \"abcdel\"): " (del-del  "abcdel"))
  (println "(del-del \"add\"): " (del-del  "add"))
  (println "(del-del \"ad\"): " (del-del  "ad"))
  (println "(del-del \"a\"): " (del-del  "a"))
  (println "(del-del \"\"): " (del-del  ""))
  (println "(del-del \"del\"): " (del-del  "del"))
  (println "(del-del \"adel\"): " (del-del  "adel"))
  (println "(del-del \"aadelbb\"): " (del-del  "aadelbb"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (mix-start)")
  (println "(mix-start \"mix snacks\"): " (mix-start  "mix snacks"))
  (println "(mix-start \"pix snacks\"): " (mix-start  "pix snacks"))
  (println "(mix-start \"piz snacks\"): " (mix-start  "piz snacks"))
  (println "(mix-start \"nix\"): " (mix-start  "nix"))
  (println "(mix-start \"8ix\"): " (mix-start  "8ix"))
  (println "(mix-start \"ni\"): " (mix-start  "ni"))
  (println "(mix-start \"n\"): " (mix-start  "n"))
  (println "(mix-start \"\"): " (mix-start  ""))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (start-oz)")
  (println "(start-oz \"ozymandias\"): " (start-oz "ozymandias"))
  (println "(start-oz \"bzoo\"): " (start-oz "bzoo"))
  (println "(start-oz \"oxx\"): " (start-oz "oxx"))
  (println "(start-oz \"oz\"): " (start-oz "oz"))
  (println "(start-oz \"ounce\"): " (start-oz "ounce"))
  (println "(start-oz \"o\"): " (start-oz "o"))
  (println "(start-oz \"abc\"): " (start-oz "abc"))
  (println "(start-oz \"\"): " (start-oz ""))
  (println "(start-oz \"zoo\"): " (start-oz "zoo"))
  (println "(start-oz \"aztec\"): " (start-oz "aztec"))
  (println "(start-oz \"zzzz\"): " (start-oz "zzzz"))
  (println "(start-oz \"oznic\"): " (start-oz "oznic"))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (int-max)")
  (println "(int-max 1 2 3)): " (int-max 1 2 3))
  (println "(int-max 1 3 2)): " (int-max 1 3 2))
  (println "(int-max 3 2 1)): " (int-max 3 2 1))
  (println "(int-max 9 3 3)): " (int-max 9 3 3))
  (println "(int-max 3 9 3)): " (int-max 3 9 3))
  (println "(int-max 3 3 9)): " (int-max 3 3 9))
  (println "(int-max 8 2 3)): " (int-max 8 2 3))
  (println "(int-max -3 -1 -2)): " (int-max -3 -1 -2))
  (println "(int-max 6 2 5)): " (int-max 6 2 5))
  (println "(int-max 5 6 2)): " (int-max 5 6 2))
  (println "(int-max 5 2 6)): " (int-max 5 2 6))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (close10)")
  (println "(close10 8 13)): " (close10 8 13))
  (println "(close10 13 8)): " (close10 13 8))
  (println "(close10 13 7)): " (close10 13 7))
  (println "(close10 7 13)): " (close10 7 13))
  (println "(close10 9 13)): " (close10 9 13))
  (println "(close10 13 8)): " (close10 13 8))
  (println "(close10 10 12)): " (close10 10 12))
  (println "(close10 11 10)): " (close10 11 10))
  (println "(close10 5 21)): " (close10 5 21))
  (println "(close10 0 20)): " (close10 0 20))
  (println "(close10 10 10)): " (close10 10 10))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing in3050()")
  (println "(in3050 30 31)): " (in3050 30 31))
  (println "(in3050 30 41)): " (in3050 30 41))
  (println "(in3050 40 50)): " (in3050 40 50))
  (println "(in3050 40 51)): " (in3050 40 51))
  (println "(in3050 39 50)): " (in3050 39 50))
  (println "(in3050 50 39)): " (in3050 50 39))
  (println "(in3050 40 39)): " (in3050 40 39))
  (println "(in3050 49 48)): " (in3050 49 48))
  (println "(in3050 50 40)): " (in3050 50 40))
  (println "(in3050 50 51)): " (in3050 50 51))
  (println "(in3050 35 36)): " (in3050 35 36))
  (println "(in3050 35 45)): " (in3050 35 45))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing max1020()")
  (println "(max1020 11 19)): " (max1020 11 19))
  (println "(max1020 19 11)): " (max1020 19 11))
  (println "(max1020 11 9)): " (max1020 11 9))
  (println "(max1020 9 21)): " (max1020 9 21))
  (println "(max1020 10 21)): " (max1020 10 21))
  (println "(max1020 21 10)): " (max1020 21 10))
  (println "(max1020 9 11)): " (max1020 9 11))
  (println "(max1020 23 10)): " (max1020 23 10))
  (println "(max1020 20 10)): " (max1020 20 10))
  (println "(max1020 7 20)): " (max1020 7 20))
  (println "(max1020 17 16)): " (max1020 17 16))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (string-e)")
  (println "(string-e \"Hello\")): " (string-e  "Hello"))
  (println "(string-e \"Heelle\")): " (string-e  "Heelle"))
  (println "(string-e \"Heelele\")): " (string-e  "Heelele"))
  (println "(string-e \"Hll\")): " (string-e  "Hll"))
  (println "(string-e \"e\")): " (string-e  "e"))
  (println "(string-e \"\")): " (string-e  ""))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (last-digit)")
  (println "(last-digit 7 17)): " (last-digit 7 17))
  (println "(last-digit 6 17)): " (last-digit 6 17))
  (println "(last-digit 3 113)): " (last-digit 3 113))
  (println "(last-digit 114 113)): " (last-digit 114 113))
  (println "(last-digit 114 4)): " (last-digit 114 4))
  (println "(last-digit 10 0)): " (last-digit 10 0))
  (println "(last-digit 11 0)): " (last-digit 11 0))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (end-up)")
  (println "(end-up \"Hello\")): " (end-up "Hello"))
  (println "(end-up \"hi there\")): " (end-up "hi there"))
  (println "(end-up \"hi\")): " (end-up "hi"))
  (println "(end-up \"woo hoo\")): " (end-up "woo hoo"))
  (println "(end-up \"xyz12\")): " (end-up "xyz12"))
  (println "(end-up \"x\")): " (end-up "x"))
  (println "(end-up \"\")): " (end-up ""))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (every-nth)")
  (println "(every-nth \"Miracle\" 2)): " (every-nth "Miracle" 2))
  (println "(every-nth \"abcdefg\" 2)): " (every-nth "abcdefg" 2))
  (println "(every-nth \"abcdefg\" 3)): " (every-nth "abcdefg" 3))
  (println "(every-nth \"Chocolate\" 3)): " (every-nth "Chocolate" 3))
  (println "(every-nth \"Chocolates\" 3)): " (every-nth "Chocolates" 3))
  (println "(every-nth \"Chocolates\" 4)): " (every-nth "Chocolates" 4))
  (println "(every-nth \"Chocolates\" 100)): " (every-nth "Chocolates" 100))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (string-times)")
  (println "(string-times \"Java\" 3) " (string-times "Java" 3))
  (println "(string-times \"Raleigh\" 5) " (string-times "Raleigh" 5))
  (println "(string-times \"Fun\" 7) " (string-times "Fun" 7))
  (println "(string-times \"a\" 3) " (string-times "a" 3))
  (println "(string-times \"\" 6) " (string-times "" 6))
  (println "(string-times \"program\" 4) " (string-times "program" 4))

  (println)
  (println "++++++++++++++++++++++++++++++++++++")
  (println "testing (explode)")
  (println "(explode \"example\") " (explode "example"))
  (println "(explode \"cat\") " (explode "cat"))
  (println "(explode \"c\") " (explode "c"))
  (println "(explode \"\") " (explode ""))
  (println "(explode \"craft\") " (explode "craft"))
  (println "(explode \"Capital\") " (explode "Capital"))
  (println "(explode \"home time\") " (explode "home time")))

(-main)